File: sleepy-test/csharp-xunit/ArchiveFileNameHelper_DeleteArchiveFilesByDate.cs
using Xunit;

namespace nlog
{
    public class ArchiveFileNameHelper
    {
        [Fact]
        public void DeleteArchiveFilesByDate()
        {
            {
                const int maxArchiveFiles = 3;
                var tempPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
                var logFile = Path.Combine(tempPath, "file.txt");
                try
                {
                    string archiveFolder = Path.Combine(tempPath, "archive");
                    var fileTarget = new FileTarget
                    {
                        FileName = logFile,
                        ArchiveFileName = Path.Combine(archiveFolder, "{#}.txt"),
                        ArchiveAboveSize = 50,
                        LineEnding = LineEndingMode.LF,
                        ArchiveNumbering = ArchiveNumberingMode.Date,
                        ArchiveDateFormat = "yyyyMMddHHmmssfff", //make sure the milliseconds are set in the filename
                        Layout = "${message}",
                        MaxArchiveFiles = maxArchiveFiles
                    };
                    SimpleConfigurator.ConfigureForTargetLogging(fileTarget, LogLevel.Debug);
                    //writing 19 times 10 bytes (9 char + linefeed) will result in 3 archive files and 1 current file
                    for (var i = 0; i < 19; ++i)
                    {
                        logger.Debug("123456789");
                        //build in a small sleep to make sure the current time is reflected in the filename
                        //do this every 5 entries
                        if (i % 5 == 0)
                            Thread.Sleep(50);
                    }
            
                    //Setting the Configuration to [null] will result in a 'Dump' of the current log entries
                    LogManager.Configuration = null; // Flush
                    var files = Directory.GetFiles(archiveFolder).OrderBy(s => s);
                    //the amount of archived files may not exceed the set 'MaxArchiveFiles'
                    Assert.Equal(maxArchiveFiles, files.Count());
                    SimpleConfigurator.ConfigureForTargetLogging(fileTarget, LogLevel.Debug);
                    //writing just one line of 11 bytes will trigger the cleanup of old archived files
                    //as stated by the MaxArchiveFiles property, but will only delete the oldest file
                    logger.Debug("1234567890");
                    LogManager.Configuration = null; // Flush
                    var files2 = Directory.GetFiles(archiveFolder).OrderBy(s => s);
                    Assert.Equal(maxArchiveFiles, files2.Count());
                    //the oldest file should be deleted
                    Assert.DoesNotContain(files.ElementAt(0), files2);
                    //two files should still be there
                    Assert.Equal(files.ElementAt(1), files2.ElementAt(0));
                    Assert.Equal(files.ElementAt(2), files2.ElementAt(1));
                    //one new archive file should be created
                    Assert.DoesNotContain(files2.ElementAt(2), files);
                }
                finally
                {
                    if (File.Exists(logFile))
                        File.Delete(logFile);
                    if (Directory.Exists(tempPath))
                        Directory.Delete(tempPath, true);
                }
            }
        }
    }
}

File: sleepy-test/python-unittest/ArchiveFileNameHelper_DeleteArchiveFilesByDate.py
import unittest
import os
import time
import shutil
from pathlib import Path

class TestArchiveFileNameHelper(unittest.TestCase):
    def test_delete_archive_files_by_date(self):
        max_archive_files = 3
        temp_path = os.path.join(Path(os.getenv("TEMP") or "/tmp"), str(uuid.uuid4()))
        log_file = os.path.join(temp_path, "file.txt")

        try:
            archive_folder = os.path.join(temp_path, "archive")
            file_target = FileTarget(
                file_name=log_file,
                archive_file_name=os.path.join(archive_folder, "{#}.txt"),
                archive_above_size=50,
                line_ending=LineEndingMode.LF,
                archive_numbering=ArchiveNumberingMode.DATE,
                archive_date_format="yyyyMMddHHmmssfff",
                layout="${message}",
                max_archive_files=max_archive_files,
            )

            SimpleConfigurator.configure_for_target_logging(file_target, LogLevel.DEBUG)

            for i in range(19):
                logger.debug("123456789")
                if i % 5 == 0:
                    time.sleep(0.05)

            LogManager.configuration = None
            files = sorted(Path(archive_folder).iterdir())
            self.assertEqual(max_archive_files, len(files))

            SimpleConfigurator.configure_for_target_logging(file_target, LogLevel.DEBUG)
            logger.debug("1234567890")
            LogManager.configuration = None
            files2 = sorted(Path(archive_folder).iterdir())
            self.assertEqual(max_archive_files, len(files2))
            self.assertNotIn(files[0], files2)
            self.assertEqual(files[1], files2[0])
            self.assertEqual(files[2], files2[1])
            self.assertNotIn(files2[2], files)
        finally:
            if os.path.exists(log_file):
                os.remove(log_file)
            if os.path.exists(temp_path):
                shutil.rmtree(temp_path)
